### PR TITLE
Linear Regression tutorial improvements.

### PR DESCRIPTION
**Pull request by madjam**

### CODE REVIEW

File: docs/tutorials/python/linear-regression.md [hunk 1]
```diff
@@ -30,60 +28,57 @@ eval_label = np.array([11,26,16])
 ```
 
 Once we have the data ready, we need to put it into an iterator and specify
-parameters such as the 'batch_size', and 'shuffle' which will determine the size
-of data the iterator feeds during each pass, and whether or not the data will be
-shuffled respectively.
+parameters such as `batch_size` and `shuffle`. `batch_size` determines the amount
+of data the iterator feeds during each time and `shuffle` tells whether or not to shuffle  input data.
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: "the amount of data" and "during each time" are both a bit vague. 
Please change to "the number of examples shown to the model each time we update its parameters"

Also "`shuffle` tells whether or not to shuffle" is a bit circular. Can you be more specific - does it say whetehr the data will be shuffled *ever* or whether or not it will be shuffled before each update?

File: docs/tutorials/python/linear-regression.md [hunk 2]
```diff
@@ -30,60 +28,57 @@ eval_label = np.array([11,26,16])
...
-
-2. Symbols: The actual MXNet network is defined using symbols. MXNet has
-   different types of symbols, including data placeholders, neural network
-   layers, and loss function symbols based on our requirement.
+1. **IO:** The IO class as we already saw works on the data and carries out
+   operations such as feeding data in batches and shuffling.
+   
+2. **Symbol:** The actual MXNet neural network is composed using symbols. MXNet has
+   different types of symbols, including variable placeholders for input data,
+   neural network layers and operators that manipulate NDArrays.
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: Please you the Oxford comma (place a comma before "and").

File: docs/tutorials/python/linear-regression.md [hunk 3]
```diff
@@ -30,60 +28,57 @@ eval_label = np.array([11,26,16])
...
 
-3. IO: The IO class as we already saw works on the data, and carries out
-   operations like breaking the data into batches and shuffling it.
+3. **Module:** The module class in MXNet is used to define the overall computation.
+	It is initialized with model we want to train, the training inputs (data and labels)
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: "It is initialized with model we want to train" missing "the" before model

File: docs/tutorials/python/linear-regression.md [hunk 4]
```diff
@@ -30,60 +28,57 @@ eval_label = np.array([11,26,16])
...
-   [own loss][https://github.com/dmlc/mxnet/blob/master/docs/tutorials/r/CustomLossFunction.md#how-to-use-your-own-loss-function]
-   Some examples of existing losses are: LinearRegressionOutput, which computes
-   the l2-loss between it's input symbol and the actual labels provided to it,
-   SoftmaxOutput, which computs the categorical cross-entropy.
-
-The ones described above, and other symbols are chained one after the other,
-servng as input to one another to create the network topology. More information
-about the different types of symbols can be found
-[here][http://mxnet.io/api/python/symbol.html]
+   when we are commence training.
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: remove "are"

File: docs/tutorials/python/linear-regression.md [hunk 5]
```diff
@@ -94,32 +89,31 @@ lro = mx.sym.LinearRegressionOutput(data=fully_connected_layer, label=Y, name="l
...
-1. FullyConnected: The fully connected symbol represents a fully connected layer
+1. `FullyConnected`: The fully connected symbol represents a fully connected layer
    of a neural network (without any activation being applied), which in essence,
    is just a linear regression on the input attributes. It takes the following
    parameters:
 
-   - data: Input to the layer (specify the symbol whose output should be fed here)
-   - num_hidden: Number of hidden dimension which specifies the size of the output of the layer
+   - `data`: Input to the layer (specifies the symbol whose output should be fed here)
+   - `num_hidden`: Number of hidden dimensions which specifies the size of the output of the layer
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: Change to 
"The number of hidden neurons in the layer's output."

File: docs/tutorials/python/linear-regression.md [hunk 6]
```diff
@@ -94,32 +89,31 @@ lro = mx.sym.LinearRegressionOutput(data=fully_connected_layer, label=Y, name="l
...
 
-2. Linear Regression Output: Output layers in MXNet aim at implementing a
-   loss. In our example, the Linear Regression Output layer is used which
-   specifies that an l2 loss needs to be applied against it's input and the
-   actual labels provided to this layer. The parameters to this layer are:
+2. `LinearRegressionOutput`: Output layers in MXNet aim to implement a
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: "Output layers in MXNet aim to implement a loss"

Layers don't aim to do anything. Please reword. What precisely is meant to be said here?

File: docs/tutorials/python/linear-regression.md [hunk 7]
```diff
@@ -94,32 +89,31 @@ lro = mx.sym.LinearRegressionOutput(data=fully_connected_layer, label=Y, name="l
...
+   loss. In our example, the `LinearRegressionOutput` layer specifies that an l2 loss
+   needs to be applied against it's input and the labels provided to this layer. The 	parameters to this layer are:
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: "it's" -> "its"

File: docs/tutorials/python/linear-regression.md [hunk 8]
```diff
@@ -94,32 +89,31 @@ lro = mx.sym.LinearRegressionOutput(data=fully_connected_layer, label=Y, name="l
...
-train_iter = mx.io.NDArrayIter(..., label_name="lin_reg_label")**
+**Note on naming convention:** the label variable's name should be the same as the
+`label_name` parameter passed to your training data iterator. The default value of
+this is `softmax_label`, but we have updated it to `lin_reg_label` in this
+tutorial as you can see in `Y = mx.symbol.Variable('lin_reg_label')` and
+`train_iter = mx.io.NDArrayIter(..., label_name='lin_reg_label')`.
 
-Finally, the network is stored into a *Module*, where you define the symbol
-who's value is to be minimised (in our case, lro or the lin_reg_output"), the
+Finally, the network is input to a *Module*, where you define the symbol
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: Let's (almost) never use the pronoun "you". Instead, favor the scientific usage of "we" meaning the authors and reader, together. 

Change to:
"Finally, we give the network as input to a *Module*, where we define..."

File: docs/tutorials/python/linear-regression.md [hunk 9]
```diff
@@ -94,32 +89,31 @@ lro = mx.sym.LinearRegressionOutput(data=fully_connected_layer, label=Y, name="l
...
+the needs to be minimized (in our case, `lro` or the `lin_reg_output`), the
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: "the needs to be min" -> "*that* needs to be..."

File: docs/tutorials/python/linear-regression.md [hunk 10]
```diff
@@ -138,8 +132,8 @@ mx.viz.plot_network(symbol=lro)
 ## Training the model
 
 Once we have defined the model structure, the next step is to train the
-parameters of the model to fit the training data. This is done by using the
-**fit()** function of the **Module** class.
+parameters of the model to fit the training data. This is done using the
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: change "done" to "accomplished"

File: docs/tutorials/python/linear-regression.md [hunk 11]
```diff
@@ -150,21 +144,20 @@ model.fit(train_iter, eval_iter,
 
 ## Using a trained model: (Testing and Inference)
 
-Once we have a trained model, we can do multiple things on it. We can use it for inference, we can evaluate the trained model on test data. This is shown below.
+Once we have a trained model, we can do couple of things with it - we can either use it for inference or we can evaluate the trained model on test data. The later is shown below:
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: "do couple of" -> "do a couple of"

Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: "later" -> "latter"

File: docs/tutorials/python/linear-regression.md [hunk 12]
```diff
@@ -150,21 +144,20 @@ model.fit(train_iter, eval_iter,
...
 
 ```python
 model.predict(eval_iter).asnumpy()
 ```
 
-We can also evaluate our model for some metric. In this example, we are evaulating our model's mean squared error on the evaluation data.
+We can also evaluate our model for some metric. In this example, we are evaulating our model's mean squared error (MSE) on the evaluation data.
```
Comment (zackchase) [inline docs/tutorials/python/linear-regression.md]: change "for some metric" -> "according to some metric"

**Comment by madjam** [comment]
Thanks, @zackchase. Pushed updates.

File: docs/tutorials/python/linear-regression.md [hunk 13]
```diff
@@ -30,60 +28,58 @@ eval_label = np.array([11,26,16])
...
 train_iter = mx.io.NDArrayIter(train_data,train_label, batch_size, shuffle=True,label_name='lin_reg_label')
 eval_iter = mx.io.NDArrayIter(eval_data, eval_label, batch_size, shuffle=False)
 ```
 
-In the above example, we have made use of NDArrayIter, which is used to iterate
-over numpy arrays. In general, there are many different types of iterators in
-MXNet based on the type of data you will be using. Their complete documentation
-can be found at: http://mxnet.io/api/python/io.html
+In the above example, we have made use of `NDArrayIter`, which is useful for iterating
+over numpy arrays. In general, there are different types of iterators in
```
Comment (mli) [inline docs/tutorials/python/linear-regression.md]: it works for both numpy ndarray and mxnet ndarrays.

Comment (madjam) [inline docs/tutorials/python/linear-regression.md]: Fixed it.

**Comment by madjam** [comment]
@zackchase: Pushed a change to address feedback from @mli
We should merge this after it builds.

**Comment by madjam** [comment]
This PR is messed up beyond repair. Will open a new one.
